### DISCUSSION TITLE
p2ptrans generates weird geometries for simple transition

### DISCUSSION THREAD

**Discussion by samtsevich** [ftherrien/p2ptrans#27] [Q&A] [answered]
Here is the problem I faced during the usage of this script. So, I tried to generate some simple transition pathway with a small number of atoms in the unit cell and chose ZnO structures diamond-like and lonsdaleite-like ZnO. The geometries were taken from MaterialsProject and assigned as _mp-1986_ and _mp-2133_ – both structures are attached below.

So, I ran the script with default parameters, like
`p2ptrans -I mp-1986.vasp -F mp-2133.vasp -o 1986_2133` 
and the resulting trajectory is very weird – initial and final structures have totally wrong geometry (see screenshots below).

This problem is happening on any kind of system (Linux and macOS) and on any kind of script version above 2.0.0.
So, I have attached all the necessary files below and if you need some more info – let me know.

As far as I understand, this weird bug is easy to fix.

<img width="849" alt="image" src="https://user-images.githubusercontent.com/11028806/182362940-7021085b-b29d-4a51-8641-f059020670f9.png">
<img width="849" alt="image" src="https://user-images.githubusercontent.com/11028806/182363019-e499700e-fca9-4ca4-a749-cc76e11a2890.png">
[bug_p2ptrans.zip](https://github.com/ftherrien/p2ptrans/files/9241749/bug_p2ptrans.zip)

**Comment by samtsevich** [comment]
Somehow archive with inputs and outputs has not been attached properly, so I will put the link to it once again in this comment.

[`https://github.com/ftherrien/p2ptrans/files/9241749/bug_p2ptrans.zip`](https://github.com/ftherrien/p2ptrans/files/9241749/bug_p2ptrans.zip)

**Comment by ftherrien** [comment, marked as answer]
Hi @samtsevich, 

Looking at your files it seems like you ran `p2ptrans` with a smaller `-n` setting of 20, something like:

`p2ptrans -I mp-1986.vasp -F mp-2133.vasp -o 1986_2133 -n 20`

The output says the following:

```
WARNING: There are missing atoms in both structures:
Zn2 O2 (File 2) is missing 6 atoms out of 12
Zn1 O1 (File 1) is missing 6 atoms out of 12
```

This is why you are seeing weird structures where atoms are missing. This is happening because the number of atoms that you are using for mapping is smaller than the transformation cell. In other word `p2ptrans` does not have enough atoms to reconstruct the full transformation. The resulting transformation mechanism does not make physical sense.

To solve this problem you simply need to increase `-n`. As a reminder, `p2ptrans` matches spherical finite portions of the two crystals together and then finds periodicity in the mapping. `-n` controls the size of those finite portions. If you rerun with: 

 `p2ptrans -I mp-1986.vasp -F mp-2133.vasp -o 1986_2133_n100 -n 100`

You will get something like this:

![anim_n100](https://user-images.githubusercontent.com/26831168/182422680-d3ef15a4-9001-49ee-87c7-80b962e02788.gif)
F-43m (216) -> Cm (8) -> Imm2 (44) -> Cm (8) -> P6_3mc (186)

Using smaller `-n` usually gives you simpler mechanisms that involve less atoms and more strain. As you increase `-n` you will find more complex mechanisms that involve more atoms, but are more representative of the actual physical mechanism (where the number of atoms involved is very large). The downside is computational cost both for `p2ptrans` itself but also because complex mechanisms often involve larger unit cells that are slower in DFT. For example, I ran 

`p2ptrans -I mp-1986.vasp -F mp-2133.vasp -o 1986_2133_n100 -n 200`

It took about 5 min to run and I found the following:

![anim_n200](https://user-images.githubusercontent.com/26831168/182425340-9a88a6c4-51c6-4768-b2c0-56ae7d53e951.gif)
F-43m (216) -> Cc (9) -> P6_3mc (186)

With '-n 300' you get the same mechanism in about 40 min which indicates that this mechanism holds for larger sizes.

**Reply by ftherrien** [reply]
**Note 1:**

If you are limited by computation and you find yourself in a situation where you get: 
`WARNING: There are missing atoms in both structures:`

You can rerun `p2ptrans` with: 

p2ptrans -u [output folder] -cn [a number higher than -n]

This will use the transformation cell and matrix that you already found and redo the mapping using the specified size. Increase `-cn` until the warning disappears. This is not ideal, but when very large unit cells are involved, it might be the only affordable solution.

**Reply by ftherrien** [reply]
**Note 2:**

To create gifs of the transformations I used `mview` (comes with p2ptrans):

```
mview -i 1986_2133_n100/TransPOSCARS/POSCAR_0* -r 3 3 3 -showcell False -bg 250 250 250 -o anim_n100
```
Rotated the structure as necessary and waited for one full transformation cycle. Then:
```
ffmpeg -r 15 -i anim_n100_%d.png -vf scale=800:-1:flags=lanczos,palettegen palette.png
ffmpeg -r 15 -i anim_n100_%d.png -i palette.png -filter_complex "scale=800:-1:flags=lanczos[x];[x][1:v]paletteuse" anim_n100.gif
```

Sometimes mview does not want to stop so I use `Ctrl+Z` and `kill %1`
